File: Genetibase.MathX/Genetibase.MathX.NuGenCCalc_x86/src/Genetibase.MathX.NugenCCalc/EventSubscriber.cs
//using System;
//using System.Reflection;
//using System.Collections;
//using System.CodeDom;
//using System.CodeDom.Compiler;

//namespace Genetibase.MathX.NugenCCalc
//{
//    /// <summary>This delegate just for internal use.</summary>
//    public delegate void CustomEventHandler(params object[] args);

//    /// <summary>This class just for internal use.</summary>
//    public class EventSubscriber
//    {
//        private static readonly ICodeCompiler _compiler = new Microsoft.CSharp.CSharpCodeProvider().CreateCompiler();

//        private EventInfo _eventInfo;
//        private Type _compiledClass;

//        public EventSubscriber(EventInfo ei)
//        {
//            _eventInfo = ei;			
//            Compile();	
//        }

//        public void Subscribe(object obj, CustomEventHandler handler)
//        {
//            Activator.CreateInstance(_compiledClass,new object[]{obj,handler});
//        }

//        private Hashtable GetAssemblyReferencesByType(Type t)
//        {		
//            Hashtable result = new Hashtable();
//            GetAssemblyReferencesByType(t,result);
//            return result;
//        }

//        private void GetAssemblyReferencesByType(Type t, Hashtable references)
//        {
//            if (t.BaseType!=null) GetAssemblyReferencesByType(t.BaseType,references);
//            foreach (Type i in t.GetInterfaces())
//            {
//                GetAssemblyReferencesByType(i,references);
//            }
//            if (!references.ContainsKey(t.Assembly.Location))
//                references.Add(t.Assembly.Location,t.Assembly);
//        }

//        private void Compile()
//        {
//            /*
			
//            Lets generate something like that
			
//            public class ButtonClickEventSubscriber
//            {
//                CustomEventHandler _handler;
				
//                public ButtonClickEventSubscriber(object obj, CustomEventHandler handler)
//                {
//                    _handler = handler;
//                    ((button)obj).Click += new ClickEventHandler(Event);
//                }
				
//                private Event(object sender, EventArgs args)
//                {
//                    _handler(new object[]{sender,args});
//                }	
//            }
//            */

			
//            ParameterInfo[] eventParams = _eventInfo.EventHandlerType.GetMethod("Invoke").GetParameters();

//            string className = String.Format("{0}{1}EventSubscriber",_eventInfo.DeclaringType.Name,_eventInfo.Name);

//            CodeCompileUnit cu = new CodeCompileUnit();
//            CodeNamespace ns = new CodeNamespace("Genetibase.MathX.NugenCCalc.Runtime");
			
//            CodeMemberField field = new CodeMemberField(typeof(CustomEventHandler), "_handler");

//            CodeConstructor constructor = new CodeConstructor();
//            constructor.Attributes = MemberAttributes.Public;

//            CodeMemberMethod method = new CodeMemberMethod();
//            CodeTypeDeclaration cl = new CodeTypeDeclaration(className);
//            CompilerResults compile_result;

//            cu.Namespaces.Add(ns);
//            ns.Imports.Add(new CodeNamespaceImport("System"));
//            ns.Imports.Add(new CodeNamespaceImport("Genetibase.MathX.NugenCCalc"));

//            ns.Types.Add(cl);

//            cl.TypeAttributes = TypeAttributes.Public;
//            cl.Members.Add(field);
//            cl.Members.Add(method);
//            cl.Members.Add(constructor);
			
//            cu.ReferencedAssemblies.Add(this.GetType().Assembly.Location);

//            foreach(object o in GetAssemblyReferencesByType(_eventInfo.DeclaringType).Keys)
//            {
//                cu.ReferencedAssemblies.Add((string)o);
//            }
			
//            // Constructor

//            constructor.Attributes = MemberAttributes.Public;			
//            constructor.Parameters.Add(
//                new CodeParameterDeclarationExpression(typeof(object),"obj"));
//            constructor.Parameters.Add(
//                new CodeParameterDeclarationExpression(typeof(CustomEventHandler),"handler"));
//            constructor.Statements.Clear();

//            constructor.Statements.Add(new CodeExpressionStatement(
//                new CodeSnippetExpression("_handler = handler")));
//            constructor.Statements.Add(new CodeExpressionStatement(
//                new CodeSnippetExpression(
//                string.Format("(({0})obj).{1} += new {2}(Event)",_eventInfo.DeclaringType.FullName,_eventInfo.Name,_eventInfo.EventHandlerType.FullName)
//                )));

//            // Event method

//            method.ReturnType = null;
//            method.Attributes = MemberAttributes.Public;
//            method.Name="Event";

//            ArrayList paramNames = new ArrayList();
			
//            foreach (ParameterInfo pi in eventParams)
//            {
//                method.Parameters.Add(new CodeParameterDeclarationExpression(pi.ParameterType,pi.Name));
//                paramNames.Add (pi.Name);
//            }

//            method.Statements.Clear();
//            method.Statements.Add(new CodeExpressionStatement(
//                new CodeSnippetExpression(
//                String.Format("_handler({0})",string.Join(", ",(string[])paramNames.ToArray(typeof(string))))
//                )));

//            // Compile code

//            CompilerParameters compparams = new CompilerParameters();
//            compparams.GenerateInMemory=true;
			
//            compile_result = _compiler.CompileAssemblyFromDom( compparams, cu);
			
//            if ( compile_result == null || compile_result.Errors.Count > 0 )
//                throw new Exception("Cant compile");

//            _compiledClass = compile_result.CompiledAssembly.GetType(
//                "Genetibase.MathX.NugenCCalc.Runtime." + className);		
//        }

//    }
//}

File: Genetibase.MathX/Genetibase.MathX.NuGenCCalc_x86/src/Genetibase.MathX.NugenCCalc.Adapters/EventSubscriber.vb
Imports System.Reflection
Imports System.Collections
Imports System.CodeDom
Imports System.CodeDom.Compiler

Public Delegate Sub CustomEventHandler(ByVal args() As Object)

Public Class EventSubscriber

    Private Shared ReadOnly _compiler As ICodeCompiler = New Microsoft.CSharp.CSharpCodeProvider().CreateCompiler()
    Private _eventInfo As EventInfo
    Private _compiledClass As Type

    Public Sub New(ByVal ei As EventInfo)
        _eventInfo = ei
        Compile()
    End Sub

    Public Sub Subscribe(ByVal obj As Object, ByVal handler As CustomEventHandler)
        Activator.CreateInstance(_compiledClass, New Object() {obj, handler})
    End Sub

    Private Function GetAssemblyReferencesByType(ByVal t As Type) As Hashtable
        Dim result As Hashtable = New Hashtable
        GetAssemblyReferencesByType(t, result)
        Return result
    End Function

    Private Sub GetAssemblyReferencesByType(ByVal t As Type, ByVal references As Hashtable)
        If (Not t.BaseType Is Nothing) Then
            GetAssemblyReferencesByType(t.BaseType, references)
        End If
        For Each i As Type In t.GetInterfaces()
            GetAssemblyReferencesByType(i, references)
        Next
        If (Not references.ContainsKey(t.Assembly.Location)) Then
            references.Add(t.Assembly.Location, t.Assembly)
        End If
    End Sub

    Private Sub Compile()
        Dim eventParams() As ParameterInfo = _eventInfo.EventHandlerType.GetMethod("Invoke").GetParameters()
        Dim className As String = String.Format("{0}{1}EventSubscriber", _eventInfo.DeclaringType.Name, _eventInfo.Name)

        Dim cu As CodeCompileUnit = New CodeCompileUnit()
        Dim ns As CodeNamespace = New CodeNamespace("Genetibase.MathX.NugenCCalc.Runtime")

        Dim field As CodeMemberField = New CodeMemberField(GetType(CustomEventHandler), "_handler")

        Dim constructor As CodeConstructor = New CodeConstructor()
        constructor.Attributes = MemberAttributes.Public

        Dim method As CodeMemberMethod = New CodeMemberMethod()
        Dim cl As CodeTypeDeclaration = New CodeTypeDeclaration(className)
        Dim compile_result As CompilerResults

        cu.Namespaces.Add(ns)
        ns.Imports.Add(New CodeNamespaceImport("System"))
        ns.Imports.Add(New CodeNamespaceImport("Genetibase.MathX.NugenCCalc.Adapters"))

        ns.Types.Add(cl)

        cl.TypeAttributes = TypeAttributes.Public
        cl.Members.Add(field)
        cl.Members.Add(method)
        cl.Members.Add(constructor)

        cu.ReferencedAssemblies.Add(Me.GetType().Assembly.Location)

        For Each o As Object In GetAssemblyReferencesByType(_eventInfo.DeclaringType).Keys
            cu.ReferencedAssemblies.Add(o.ToString())
        Next

        constructor.Attributes = MemberAttributes.Public
        constructor.Parameters.Add(New CodeParameterDeclarationExpression(GetType(Object), "obj"))
        constructor.Parameters.Add(New CodeParameterDeclarationExpression(GetType(CustomEventHandler), "handler"))
        constructor.Statements.Clear()

        constructor.Statements.Add(New CodeExpressionStatement(New CodeSnippetExpression("_handler = handler")))
        constructor.Statements.Add(New CodeExpressionStatement(New CodeSnippetExpression(String.Format("(({0})obj).{1} += new {2}(Event)", _eventInfo.DeclaringType.FullName, _eventInfo.Name, _eventInfo.EventHandlerType.FullName))))

        method.ReturnType = Nothing
        method.Attributes = MemberAttributes.Public
        method.Name = "Event"

        Dim paramNames As ArrayList = New ArrayList()

        For Each pi As ParameterInfo In eventParams
            method.Parameters.Add(New CodeParameterDeclarationExpression(pi.ParameterType, pi.Name))
            paramNames.Add(pi.Name)
        Next


        method.Statements.Clear()
        Dim params As String = ""
        For Each par As String In paramNames.ToArray(GetType(String))
            params = par & ", "
        Next
        If (params.Length > 2) Then
            params = params.Substring(0, params.Length - 2)
        End If
        method.Statements.Add(New CodeExpressionStatement(New CodeSnippetExpression(String.Format("_handler({0})", params))))


        Dim compparams As CompilerParameters = New CompilerParameters()
        compparams.GenerateInMemory = True

        compile_result = _compiler.CompileAssemblyFromDom(compparams, cu)

        If (compile_result Is Nothing Or compile_result.Errors.Count > 0) Then
            Dim errorString As String = ""
            For Each err As CompilerError In compile_result.Errors
                errorString = errorString & err.ErrorText + System.Environment.NewLine
            Next

            Throw New Exception("Cant compile. Error: " & errorString)
        End If

        _compiledClass = compile_result.CompiledAssembly.GetType("Genetibase.MathX.NugenCCalc.Runtime." + className)

    End Sub


End Class
